### PR TITLE
gRPC load balancing

### LINKED ISSUE #5132: Allow gRPC Client Load Balancing
Currently, we do not allow a single validator to connect to multiple beacon node backends in a round robin fashion. Ideally, we should be able to do `--beacon-rpc-provider=BEACON_NODE_IP_1,BEACON_NODE_IP_2` from the validator client, but we have to do some gRPC setup for this to work. Thanks to @mcdee, this link on stackoverflow https://stackoverflow.com/questions/50293480/dial-multiple-ip-addresses-with-golang-grpc-client can help us achieve this. This is a great first issue for anyone wishing to contribute.

### PR DESCRIPTION
**Pull request by eklm**
**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**
It allows to specify multiple addresses of the beacon nodes like --beacon-rpc-provider=BEACON_NODE_IP_1,BEACON_NODE_IP_2 in the validator node as described int #5132.

**Which issues(s) does this PR fix?**
Fixes #5132

**Other notes for review**

### CODE REVIEW

**Comment by prestonvanloon** [comment]
I think this is a good start, however there are some caveats to client side load balancing. 

Can you add some GoDoc information on how this load balancing will work? Is it a round robin scheme? Failover?

**Comment by rauljordan** [comment]
There is also a caveat of requiring sticky sessions. In particular, the validator client asks its connected beacon node to subscribe to attestation subnets on a scheduled basis. If we do round robin requests, the next beacon node in line may not be subscribed to subnets and will fail attesting. What do you think @nisdas ? Would that be a problem with this PR?

**Comment by eklm** [comment]
Currently it's "pick first" load balancer (the default in grpc). To use round robin additional configuration for the connection is needed.

File: validator/client/multiple_endpoints_grpc_resolver.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package client
+
+import (
+	"google.golang.org/grpc/resolver"
+	"strings"
```
Comment (prestonvanloon) [inline validator/client/multiple_endpoints_grpc_resolver.go:5]:
```suggestion
	"strings"
	
	"google.golang.org/grpc/resolver"
```

**Review by prestonvanloon — COMMENTED** [review]
I am not sure that this works as intended.

I tried with this: `--beacon-rpc-provider=hostA,hostB` when hostA is offline, then it fails to connect.
```
[2020-09-04 19:32:12] FATAL validator: Could not determine if beacon chain started: could not setup beacon chain ChainStart streaming client: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp hostA: connect: connection refused"
```

Likewise, when the validator client is connected to hostA and then hostA shuts down, it does not start using hostB.

**Comment by eklm** [comment]
Hm... It works for me.
Screenshot from 2020-09-05 09-13-03
Also it switches if I start and stop the beacon-chain on the another port with --beacon-rpc-provider=localhost:4001,localhost:4000

**Comment by prestonvanloon** [comment]
Screenshot from 2020-09-10 16-00-03

It still does not work for me.

**Comment by eklm** [comment]
Oh, I see. It works for multiple IP addresses not for multiple URIs. PR patches default resolver which is used when scheme is omitted and the full URI in that case is: passthrough:///addr1,addr2. So addr2 is not a URI but address of a machine.
--beacon-rpc-provider=127.0.0.1:4000,beacon.prylabs.net:30002 switches to the second node, but it looks like it can not work by ip address and requires appropriate host header but the host is the full string in that case (127.0.0.1:4000,beacon.prylabs.net:30002). Unfortunately I don't see how it could be fixed because host header is defined for the whole connection not for the specific node in grpc. Still it should work fine for listing ip addresses and dns load balancing is supported out of the box if the name is resolved to multiple ips.

**Review by prestonvanloon — APPROVED** [review]

**Comment by prestonvanloon** [comment]
> Oh, I see. It works for multiple IP addresses not for multiple URIs. PR patches default resolver which is used when scheme is omitted and the full URI in that case is: passthrough:///addr1,addr2. So addr2 is not a URI but address of a machine.
> --beacon-rpc-provider=127.0.0.1:4000,beacon.prylabs.net:30002 switches to the second node, but it looks like it can not work by ip address and requires appropriate host header but the host is the full string in that case (127.0.0.1:4000,beacon.prylabs.net:30002). Unfortunately I don't see how it could be fixed because host header is defined for the whole connection not for the specific node in grpc. Still it should work fine for listing ip addresses and dns load balancing is supported out of the box if the name is resolved to multiple ips.

We can try to fix this in another PR. Regardless of this issue, your PR is a significant improvement over existing functionality. Approved. Thanks for your submission!

**Pull request merged**